docs/source/future/core/nrunner.rst

```diff
@@ -151,34 +151,68 @@ create Runnables based on the user's input from the command line::
 
-The following pseudo-code may help to illustrate that::
+A Runner will usually be tied to a specific ``kind`` of Runnable.
+That type of relationship (Runner is capable of running kind "foo"
+and Runnable is of the same kind "foo") is the expected mechanism that
+will be employed when selecting a Runner.
 
-  runnable_instance = create_runnable('exec', uri='/bin/sleep')
-  if runnable_instance.kind == 'exec':
-     runner = create_runner_exec(runnable_instance)
+A Runner can take different forms, depending on which layer one is
+interacting with.  At the lowest layer, a Runner may be a Python class
+that inherits from :class:`avocado.core.nrunner.BaseRunner`, and
+implements at least a matching constructor method, and a ``run()``
+method that should yield dictionary(ies) as result(s).
 
-A runner should produce status information on the progress of the
-execution of a runnable.  If the runnable produces interesting
-information, it should forward that along.  For instance, using the
-``exec`` runner example, it's helpful to start producing status
-that the process has been created and it's running as soon as
-possible.  These can be as simple as a sequence of::
+At a different level, a runner can take the form of an executable that
+follows the ``avocado-runner-$KIND`` naming pattern and conforms to a
+given interface/behavior, including accepting standardized command
+line arguments and producing standardized output.
+
+.. tip:: for a very basic example of the interface expected, refer to
+         ``selftests/functional/test_nrunner_interface.py`` on the
+         Avocado source code tree.
+
+Runner output
++++++++++++++
+
+A Runner should, if possible, produce status information on the
+progress of the execution of a Runnable.  While the Runner is
+executing what a Runnable describes, should it produce interesting
+information, the Runner should attempt to forward that along its
+generated status.
+
+For instance, using the ``exec`` Runner example, it's helpful to start
+producing status that the process has been created and it's running as
+soon as possible, even if no other output has been produced by the
+executable itself.  These can be as simple as a sequence of::
 
   {"status": "running"}
   {"status": "running"}
 
-When the process is finished, it can return::
+When the process is finished, the Runner may return::
 
   {"status": "finished", "returncode": 0, 'stdout': b'', 'stderr': b''}
 
-Note that, besides the status of ``finished``, and a return code which
-can be used to determine a success or failure status, it's not the
-runner's responsibility to determine test results.
+.. tip:: Besides the status of ``finished``, and a return code which
+         can be used to determine a success or failure status, a
+         Runner may not be obliged to determine the overall PASS/FAIL
+         outcome.  Whoever called the runner may be responsible to
+         determine its overall result, including a PASS/FAIL
+         judgement.
```

Comment:
Trying to understand here...

On `avocado/core/nrunner.py` there are few examples where the Runners are evaluating the status based on return code. Ex:

```python
    def run(self):
        for status in super(ExecTestRunner, self).run():
            if 'returncode' in status:
                if status['returncode'] == 0:
                    status['status'] = 'pass'
                else:
                    status['status'] = 'fail'
            yield status
```

 Whoever called the runner will be doing this, soon? Or this will continue on Runner code?

Reply:
What I've tried to say is that the authoritative and definitive responsibility lies with "whoever called the runner" (usually the job).  This doesn't mean that a Runner can not set a clear pass/fail "signal".

The better example is the `ExecRunner` code, in which no PASS/FAIL condition is clear... suppose that we built a specialized sysinfo-like feature based on `ExecRunner` output, and we know that a given binary to be used to collect results will return:

* `0` for "no progress has happened since last probe"
* `>= 1` for "a given number of progress units have happened since the last probe"
* `<= 1` for an error condition

It'd be up to the sysinfo-like feature to decide what was a PASS/FAIL outcome.